calyx-opt/src/passes/compile_sync_without_sync_reg.rs

```diff
@@ -0,0 +1,211 @@
+use crate::traversal::{Action, Named, VisResult, Visitor};
+use calyx_ir::{self as ir, GetAttributes, RRC};
+use calyx_ir::{build_assignments, guard, structure};
+use calyx_utils::{CalyxResult, Error};
+use std::collections::HashMap;
+
+#[derive(Default)]
+/// Compiles @sync without use of std_sync_reg
+/// Upon encountering @sync, it first instantiates a std_reg(1) for each thread(`bar`)
+/// and a std_reg(1) for each barrier (`s`)
+/// It then continuously assigns the value of (`s.in`) to 1'd1 guarded by the
+/// expression that all values of `bar` for threads under the barrier are
+/// set to 1'd1
+/// Then it replaces the @sync control operator with
+/// seq {
+///     barrier;
+///     clear;
+/// }
+/// `barrier` simply sets the value of `bar` to 1'd1 and then waits
+/// for `s.out` to be up
+/// `clear` resets the value of `bar` to 1'd0 for reuse of barrier
+/// Using this method, each thread only incurs 3 cycles of latency overhead for
+/// the barrier, and we theoretically won't have a limit for number of threads
+/// under one barrier
+pub struct CompileSyncWithoutSyncReg;
+
+impl Named for CompileSyncWithoutSyncReg {
+    fn name() -> &'static str {
+        "compile-sync-without-sync-reg"
+    }
+
+    fn description() -> &'static str {
+        "Implement barriers for statements marked with @sync attribute without
+      std_sync_ref"
+    }
+}
+
+// Data structure storing the shared `s` register and the guard accumulator
+// for guarding `s.in`
+type BarrierMap = HashMap<u64, (RRC<ir::Cell>, Box<ir::Guard<ir::Nothing>>)>;
+
+trait Gettable {
```

Comment:
Need documentation for what this trait represents. Currently it seems only `BarrierMap` implements it in which case we should [newtype it](https://doc.rust-lang.org/rust-by-example/generics/new_types.html) and implement these as methods on it

Reply:
Is there a clean way to access the traits and methods defined for `HashMap` if I am to newtype the `BarrierMap`?

Reply:
Unfortunately no. You'd have to expose them as methods on the struct but the compiler is good enough to remove the overhead of the struct after compilation.

Reply:
I ended up implementing the `Deref` and `DerefMut` traits which work.

Reply:
I personally avoid implementing those and directly exposing the methods that matter.

this is useful in the case if you want barrierMap to have additional behavior in the future which is not implementable directly on the HashMap.

Reply:
Chiming in to echo this point, generally you should avoid `Deref` and `DerefMut`
unless what you're making is a smart pointer